js/lib/Modal.js

```diff
@@ -0,0 +1,300 @@
+    var speedInBytes = (bytesReceived / (timeSinceLastUploadSpeedUpdate / 1000)).toFixed(1)
+
+    var durationLeft = ''
+    if (speedInBytes > 0) {
+      durationLeft = helpers.duration(outstanding / speedInBytes)
+    }
+
+    this._uploadRate = uploadRate
+    this._durationLeft = durationLeft
+    this._lastUploadSpeedUpdateOn = +new Date()
+    this._bytesReceivedBefore = received
+  }
+
+  var txt = this._i18n.translate('uploadProgress',
+    mbReceived, mbExpected, this._uploadRate, this._durationLeft
+  )
+  this._$modal.$label.text(txt)
+
+  var currentWidth = parseInt(this._$modal.$progress.css('width'), 10)
+  var currPercent = this._$modal.$progressBar.data('percent')
+  var self = this
+
+  // if we are going backwards (due to a restart), do not animate, but reset the width
+  // of the progress bar in one go
+  if (currPercent > progress) {
+    this._$modal.$progressBar.stop().css('width', progress + '%')
+    this._$modal.$progressBar.data('percent', progress)
+    this._setProgressbarPercent(progress)
+    return
+  }
+
+  this._$modal.$progressBar.data('percent', progress)
+  this._$modal.$progressBar.stop().animate(
+    {width: progress + '%'},
+    {
+      duration: 1000,
+      easing: 'linear',
+      progress: function (promise, currPercent, remainingMs) {
+        // self._$modal can actually be gone if cancel was hit in the meantime
+        if (!self._$modal) {
+          return
+        }
+        var percent = self._setProgressbarPercent(currentWidth)
+
+        if (percent == 100 && !self._animatedTo100) {
+          self._animatedTo100 = true
+          setTimeout(function () {
+            // self._$modal can actually be gone if cancel was hit in the meantime
+            if (!self._$modal) {
+              return
+            }
+            self._renderProcessingFiles()
+          }, 500)
+        }
+      }
+    }
+  )
+}
+
+Modal.prototype._renderProcessingFiles = function () {
+  this._$modal.$label.text(this._i18n.translate('processingFiles'))
+  this._$modal.$progress.addClass('active')
+  this._$modal.$percent.text('')
+}
+
+Modal.prototype._setProgressbarPercent = function (totalWidth) {
+  var width = parseInt(this._$modal.$progressBar.css('width'), 10)
+
+  var percent = (width * 100 / totalWidth).toFixed(0)
+  if (percent > 100) {
+    percent = 100
+  }
+  if (percent > 13 && !this._animatedTo100) {
+    this._$modal.$percent.text(percent + '%')
+  }
+  if (percent <= 13) {
+    this._$modal.$percent.text('')
+  }
+
+  return percent
+}
+
+Modal.prototype._toggleErrorTexts = function (mode) {
+  if (!this._$modal) {
+    return
+  }
+
+  this._$modal.$error.toggle(mode)
+  this._$modal.$content.toggleClass('content-error', mode)
+
+  // error details have their own toggle-show mechanism
+  if (!mode) {
+    this._$modal.$errorDetails.toggle(mode)
+    this._$modal.$errorDetailsToggle.toggle(mode)
+  }
+}
+
+Modal.prototype._toggleProgressTexts = function (mode) {
+  if (!this._$modal) {
```

Comment:
This check seems to occur frequently is it possible to wrap all the functions around a checker before they are called? Maybe something like this http://javascript.info/tutorial/decorators?